WordPress/gutenberg
# Button: always render the Tooltip component even when a tooltip should not be shown

Fixes https://github.com/WordPress/gutenberg/issues/56522

Requires #57202, #57878, #57975, #58125

## What?


This PR tweaks the `Button` component so that it _always_ renders the `Tooltip` component, even when a tooltip shouldn't be shown.

It also un-does the temporary fix from https://github.com/WordPress/gutenberg/pull/56502

## Why?


This change is necessary to fix the issue described in https://github.com/WordPress/gutenberg/discussions/44735, where basically the element rendered by `Button` would be removed and re-added to the DOM any time the `Button` component switches from having to show a tooltip to not having to show it, and vice-versa.

The fact that a new HTML element is re-created every time triggers a focus loss that can be very annoying, especially to assistive technology users.

## How?


After some debugging, the issue seems related to the fact that React couldn't reconciliate the "with tooltip" and "without tooltip" version of the `Button` component without re-creating the whole button every time.

To solve this, I decided to tweak `Button` so that it always renders the `Tooltip` component.

_Note: this doesn't mean that the tooltip will always be shown, though — the original user-facing behaviour should have been retained by only passing props to the `Tooltip` component when needed._

## Testing Instructions





### Testing Instructions for Keyboard


From https://github.com/WordPress/gutenberg/discussions/44735#discussioncomment-7639716: 

- Edit a draft post and make sure it's saved so that the Save button shows the text 'Saved'.
- Make any edit so that the button text changes to 'Save draft'.
- Select the Save button in the dev tools inspector.
- Right click and enable 'Break on > node removal'.
- Click the Save button.
- Observe the button doesn't gets removed from the DOM, and the keyboard focus stays on the button

A similar approach can also be used directly in Storybook.

Also, make sure that in general, tooltips show on `Button` when they are supposed to, and don't show when they are not supposed to.
